File: app/src/main/java/it/ncorti/emgvisualizer/myo/ByteReader.java
/* This file is part of EmgVisualizer.

    EmgVisualizer is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    EmgVisualizer is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with EmgVisualizer.  If not, see <http://www.gnu.org/licenses/>.
*/
package it.ncorti.emgvisualizer.myo;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * This class help you to read the byte line from Myo.
 * But be carefully to byte array size. There is no limitation of get() method,
 * so there is a possibilty of overloading the byte buffer.
 * <p>
 * ByteReader is useful for handling raw data taken from bluetooth connection with Myo
 *
 * @author Nicola
 */
public class ByteReader {

    /**
     * Raw byte array
     */
    private byte[] byteData;
    /**
     * ByteBuffer for reading purpose
     */
    private ByteBuffer bbf;

    /**
     * Method for setting byteData into reader
     *
     * @param data Raw byteData read
     */
    public void setByteData(byte[] data) {
        this.byteData = data;
        this.bbf = ByteBuffer.wrap(this.byteData);
        bbf.order(ByteOrder.LITTLE_ENDIAN);
    }

    /**
     * Return reference to byteData
     *
     * @return byteData reference
     */
    public byte[] getByteData() {
        return byteData;
    }

    /**
     * Return a short from byteReader
     * ATTENTION: don't call this method before setByteData
     * ATTENTION: pay attention to byte array size
     *
     * @return Next short read
     */
    public short getShort() {
        return this.bbf.getShort();
    }

    /**
     * Return a byte from byteReader
     * ATTENTION: don't call this method before setByteData
     * ATTENTION: pay attention to byte array size
     *
     * @return Next byte read
     */
    public byte getByte() {
        return this.bbf.get();
    }

    /**
     * Return a int from byteReader
     * ATTENTION: don't call this method before setByteData
     * ATTENTION: pay attention to byte array size
     *
     * @return Next int read
     */
    public int getInt() {
        return this.bbf.getInt();
    }

    /**
     * Rewind byte reader to begin, for restart reading
     */
    public void rewind() {
        this.bbf.rewind();
    }

    /**
     * Method for reading n consecutive floats, returned in a new array
     *
     * @param size Number of bytes to be read (usually 8 or 16)
     * @return A new array with read bytes
     */
    public float[] getBytes(int size) {
        float[] result = new float[size];
        for (int i = 0; i < size; i++)
            result[i] = bbf.get();
        return result;
    }
}

File: myonnaise/src/main/java/com/ncorti/myonnaise/ByteReader.kt
package com.ncorti.myonnaise

import android.bluetooth.BluetoothGattCharacteristic
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * This class help you to read the byte line from Myo.
 * Please pay attention that there are no checks for BufferOverFlow or similar problems,
 * you just receive the amount of data you request (1, 2 or 4 bytes).
 *
 * [ByteReader] is useful for handling raw data taken from bluetooth connection with Myo.
 * Use the [ByteReader.getBytes] to get an array of float from a [BluetoothGattCharacteristic].
 */
class ByteReader {

    private var byteBuffer: ByteBuffer? = null

    var byteData: ByteArray? = null
        set(data) {
            field = data
            this.byteBuffer = field?.let { ByteBuffer.wrap(it) }?.apply {
                order(ByteOrder.nativeOrder())
            }
        }

    val short: Short
        get() = this.byteBuffer!!.short

    val byte: Byte
        get() = this.byteBuffer!!.get()

    val int: Int
        get() = this.byteBuffer!!.int

    fun rewind() = this.byteBuffer?.rewind()

    /**
     * Method for reading n consecutive floats, returned in a new array.
     *
     * @param size Number of bytes to be read (usually 8 or 16)
     * @return A new array with read bytes
     */
    fun getBytes(size: Int): FloatArray {
        val result = FloatArray(size)
        for (i in 0 until size)
            result[i] = byteBuffer!!.get().toFloat()
        return result
    }
}
